File: plugins/claude-swarm/agent-sdk/python/message_broker.py
"""
Claude Swarm - Message Broker
Sistema de mensageria baseado em Redis para comunicacao entre agentes.
"""

import asyncio
import json
import uuid
from datetime import datetime
from typing import Optional, Callable, Any, Dict, List
from dataclasses import dataclass, field, asdict
from enum import Enum

import redis.asyncio as redis

from config import config


class MessageType(str, Enum):
    """Tipos de mensagem do Swarm."""
    TASK = "TASK"
    RESULT = "RESULT"
    BROADCAST = "BROADCAST"
    STATUS = "STATUS"
    HEARTBEAT = "HEARTBEAT"
    ERROR = "ERROR"


class Priority(str, Enum):
    """Prioridades de mensagem."""
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


@dataclass
class Message:
    """Estrutura de mensagem do Swarm."""
    type: MessageType
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    timestamp: str = field(default_factory=lambda: datetime.utcnow().isoformat() + "Z")
    from_agent: str = field(default_factory=lambda: config.agent_id)
    to_agent: str = ""
    payload: Dict[str, Any] = field(default_factory=dict)
    metadata: Dict[str, Any] = field(default_factory=dict)

    def __post_init__(self):
        if not self.metadata:
            self.metadata = {
                "priority": Priority.MEDIUM.value,
                "ttl": 300000  # 5 minutos
            }

    def to_json(self) -> str:
        """Serializa mensagem para JSON."""
        data = {
            "type": self.type.value if isinstance(self.type, MessageType) else self.type,
            "id": self.id,
            "timestamp": self.timestamp,
            "from": self.from_agent,
            "to": self.to_agent,
            "payload": self.payload,
            "metadata": self.metadata
        }
        return json.dumps(data)

    @classmethod
    def from_json(cls, data: str) -> "Message":
        """Deserializa mensagem de JSON."""
        parsed = json.loads(data)
        return cls(
            type=MessageType(parsed.get("type", "TASK")),
            id=parsed.get("id", str(uuid.uuid4())),
            timestamp=parsed.get("timestamp", datetime.utcnow().isoformat() + "Z"),
            from_agent=parsed.get("from", "unknown"),
            to_agent=parsed.get("to", ""),
            payload=parsed.get("payload", {}),
            metadata=parsed.get("metadata", {})
        )


class MessageBroker:
    """
    Message Broker para comunicacao entre agentes.
    Usa Redis Pub/Sub para comunicacao em tempo real.
    """

    def __init__(self):
        self._client: Optional[redis.Redis] = None
        self._pubsub: Optional[redis.client.PubSub] = None
        self._subscriptions: Dict[str, Callable] = {}
        self._running = False

    async def connect(self) -> None:
        """Conecta ao Redis."""
        if self._client is None:
            self._client = redis.Redis(
                host=config.redis_host,
                port=config.redis_port,
                password=config.redis_password,
                db=config.redis_db,
                decode_responses=True
            )
            # Testar conexao
            await self._client.ping()
            print(f"[BROKER] Conectado ao Redis em {config.redis_host}:{config.redis_port}")

    async def disconnect(self) -> None:
        """Desconecta do Redis."""
        self._running = False
        if self._pubsub:
            await self._pubsub.close()
            self._pubsub = None
        if self._client:
            await self._client.close()
            self._client = None
        print("[BROKER] Desconectado do Redis")

    async def publish(
        self,
        channel: str,
        message: Message,
    ) -> str:
        """
        Publica mensagem em um canal.

        Args:
            channel: Nome do canal ou worker
            message: Mensagem a publicar

        Returns:
            ID da mensagem publicada
        """
        await self.connect()

        # Determinar canal Redis
        if channel == "broadcast" or channel == "*":
            redis_channel = "swarm:broadcast"
        elif channel == "orchestrator":
            redis_channel = "swarm:results:orchestrator"
        else:
            redis_channel = f"swarm:tasks:{channel}"

        # Publicar
        await self._client.publish(redis_channel, message.to_json())

        # Salvar referencia para coleta
        await self._client.setex(
            f"swarm:pending:{message.id}",
            300,
            json.dumps({"channel": channel, "status": "pending"})
        )

        return message.id

    async def subscribe(
        self,
        channels: List[str],
        callback: Callable[[Message], Any]
    ) -> None:
        """
        Inscreve-se em canais para receber mensagens.

        Args:
            channels: Lista de canais
            callback: Funcao chamada para cada mensagem
        """
        await self.connect()

        if self._pubsub is None:
            self._pubsub = self._client.pubsub()

        # Mapear canais
        redis_channels = []
        for channel in channels:
            if channel == "broadcast":
                redis_channels.append("swarm:broadcast")
            elif channel == "tasks":
                redis_channels.append(f"swarm:tasks:{config.agent_id}")
            elif channel == "results":
                redis_channels.append(f"swarm:results:{config.agent_id}")
            else:
                redis_channels.append(f"swarm:{channel}")

        await self._pubsub.subscribe(*redis_channels)

        for ch in redis_channels:
            self._subscriptions[ch] = callback

        print(f"[BROKER] Inscrito em: {redis_channels}")

    async def listen(self) -> None:
        """Loop de escuta de mensagens."""
        if self._pubsub is None:
            raise RuntimeError("Nao inscrito em nenhum canal")

        self._running = True
        print("[BROKER] Iniciando loop de escuta...")

        async for message in self._pubsub.listen():
            if not self._running:
                break

            if message["type"] == "message":
                channel = message["channel"]
                data = message["data"]

                try:
                    msg = Message.from_json(data)
                    callback = self._subscriptions.get(channel)
                    if callback:
                        await callback(msg)
                except Exception as e:
                    print(f"[BROKER] Erro processando mensagem: {e}")

    async def collect(
        self,
        task_id: str,
        timeout_seconds: int = 30
    ) -> Optional[Message]:
        """
        Coleta resultado de uma tarefa.

        Args:
            task_id: ID da tarefa
            timeout_seconds: Timeout em segundos

        Returns:
            Mensagem de resultado ou None se timeout
        """
        await self.connect()

        result_key = f"swarm:results:{task_id}"
        start_time = asyncio.get_event_loop().time()

        while True:
            result = await self._client.get(result_key)
            if result:
                await self._client.delete(result_key)
                return Message.from_json(result)

            elapsed = asyncio.get_event_loop().time() - start_time
            if elapsed >= timeout_seconds:
                return None

            await asyncio.sleep(0.5)

    async def store_result(
        self,
        task_id: str,
        result: Dict[str, Any],
        status: str = "success"
    ) -> None:
        """
        Armazena resultado de uma tarefa.

        Args:
            task_id: ID da tarefa
            result: Resultado
            status: Status (success, partial, failed)
        """
        await self.connect()

        message = Message(
            type=MessageType.RESULT,
            id=task_id,
            to_agent="orchestrator",
            payload={"status": status, "result": result}
        )

        result_key = f"swarm:results:{task_id}"
        await self._client.setex(result_key, 300, message.to_json())

        # Notificar orchestrator
        await self._client.publish("swarm:results:orchestrator", message.to_json())

        # Remover de pendentes
        await self._client.delete(f"swarm:pending:{task_id}")

    async def broadcast(
        self,
        action: str,
        message: str = ""
    ) -> str:
        """
        Envia broadcast para todos os workers.

        Args:
            action: Acao (pause, resume, shutdown, etc)
            message: Mensagem adicional

        Returns:
            ID do broadcast
        """
        msg = Message(
            type=MessageType.BROADCAST,
            to_agent="*",
            payload={"action": action, "message": message},
            metadata={"priority": Priority.HIGH.value, "ttl": 60000}
        )

        await self.publish("broadcast", msg)
        return msg.id

    async def heartbeat(self) -> None:
        """Envia heartbeat para registro de atividade."""
        await self.connect()

        key = f"swarm:heartbeat:{config.agent_id}"
        timestamp = datetime.utcnow().timestamp()
        await self._client.setex(key, config.heartbeat_ttl, str(timestamp))

    async def health_check(self) -> Dict[str, Any]:
        """
        Verifica saude de todos os workers.

        Returns:
            Status de cada worker
        """
        await self.connect()

        workers = []
        keys = await self._client.keys("swarm:heartbeat:*")
        now = datetime.utcnow().timestamp()

        for key in keys:
            worker_name = key.replace("swarm:heartbeat:", "")
            last_seen = await self._client.get(key)

            if last_seen:
                age_seconds = int(now - float(last_seen))
                status = "alive" if age_seconds < 30 else "dead"
            else:
                age_seconds = -1
                status = "unknown"

            workers.append({
                "name": worker_name,
                "status": status,
                "last_seen_seconds_ago": age_seconds
            })

        return {
            "workers": workers,
            "total": len(workers),
            "healthy": sum(1 for w in workers if w["status"] == "alive"),
            "unhealthy": sum(1 for w in workers if w["status"] != "alive")
        }

    async def set_state(
        self,
        key: str,
        value: Any,
        ttl_seconds: Optional[int] = None
    ) -> None:
        """
        Salva estado compartilhado.

        Args:
            key: Chave
            value: Valor
            ttl_seconds: TTL opcional
        """
        await self.connect()

        full_key = f"swarm:state:{key}"
        value_str = json.dumps(value) if not isinstance(value, str) else value

        if ttl_seconds:
            await self._client.setex(full_key, ttl_seconds, value_str)
        else:
            await self._client.set(full_key, value_str)

    async def get_state(self, key: str) -> Optional[Any]:
        """
        Recupera estado compartilhado.

        Args:
            key: Chave

        Returns:
            Valor ou None
        """
        await self.connect()

        full_key = f"swarm:state:{key}"
        value = await self._client.get(full_key)

        if value:
            try:
                return json.loads(value)
            except json.JSONDecodeError:
                return value
        return None


# Instancia global
broker = MessageBroker()

File: plugins/claude-swarm/agent-sdk/typescript/src/message-broker.ts
/**
 * Claude Swarm - Message Broker
 * Sistema de mensageria baseado em Redis para comunicacao entre agentes.
 */

import Redis from 'ioredis';
import { v4 as uuidv4 } from 'uuid';
import { config } from './config';
import { Message, MessageType, Priority, HealthCheckResult } from './types';

export class MessageBroker {
  private client: Redis | null = null;
  private subscriber: Redis | null = null;
  private subscriptions: Map<string, (message: Message) => void> = new Map();
  private running = false;

  async connect(): Promise<void> {
    if (this.client) return;

    const redisConfig = {
      host: config.redisHost,
      port: config.redisPort,
      password: config.redisPassword,
      db: config.redisDb,
    };

    this.client = new Redis(redisConfig);
    this.subscriber = new Redis(redisConfig);

    await this.client.ping();
    console.log(`[BROKER] Conectado ao Redis em ${config.redisHost}:${config.redisPort}`);
  }

  async disconnect(): Promise<void> {
    this.running = false;
    if (this.subscriber) {
      await this.subscriber.quit();
      this.subscriber = null;
    }
    if (this.client) {
      await this.client.quit();
      this.client = null;
    }
    console.log('[BROKER] Desconectado do Redis');
  }

  createMessage(
    type: MessageType,
    to: string,
    payload: Record<string, unknown>,
    priority: Priority = Priority.MEDIUM
  ): Message {
    return {
      type,
      id: uuidv4(),
      timestamp: new Date().toISOString(),
      from: config.agentId,
      to,
      payload,
      metadata: {
        priority,
        ttl: 300000, // 5 minutos
      },
    };
  }

  async publish(channel: string, message: Message): Promise<string> {
    await this.connect();

    // Determinar canal Redis
    let redisChannel: string;
    if (channel === 'broadcast' || channel === '*') {
      redisChannel = 'swarm:broadcast';
    } else if (channel === 'orchestrator') {
      redisChannel = 'swarm:results:orchestrator';
    } else {
      redisChannel = `swarm:tasks:${channel}`;
    }

    // Publicar
    await this.client!.publish(redisChannel, JSON.stringify(message));

    // Salvar referencia para coleta
    await this.client!.setex(
      `swarm:pending:${message.id}`,
      300,
      JSON.stringify({ channel, status: 'pending' })
    );

    return message.id;
  }

  async subscribe(
    channels: string[],
    callback: (message: Message) => void
  ): Promise<void> {
    await this.connect();

    // Mapear canais
    const redisChannels = channels.map((channel) => {
      if (channel === 'broadcast') return 'swarm:broadcast';
      if (channel === 'tasks') return `swarm:tasks:${config.agentId}`;
      if (channel === 'results') return `swarm:results:${config.agentId}`;
      return `swarm:${channel}`;
    });

    for (const ch of redisChannels) {
      this.subscriptions.set(ch, callback);
    }

    await this.subscriber!.subscribe(...redisChannels);
    console.log(`[BROKER] Inscrito em: ${redisChannels.join(', ')}`);
  }

  async listen(): Promise<void> {
    if (!this.subscriber) {
      throw new Error('Nao inscrito em nenhum canal');
    }

    this.running = true;
    console.log('[BROKER] Iniciando loop de escuta...');

    this.subscriber.on('message', (channel, data) => {
      if (!this.running) return;

      try {
        const message: Message = JSON.parse(data);
        const callback = this.subscriptions.get(channel);
        if (callback) {
          callback(message);
        }
      } catch (e) {
        console.error(`[BROKER] Erro processando mensagem: ${e}`);
      }
    });
  }

  async collect(taskId: string, timeoutSeconds: number = 30): Promise<Message | null> {
    await this.connect();

    const resultKey = `swarm:results:${taskId}`;
    const startTime = Date.now();

    while (true) {
      const result = await this.client!.get(resultKey);
      if (result) {
        await this.client!.del(resultKey);
        return JSON.parse(result);
      }

      const elapsed = (Date.now() - startTime) / 1000;
      if (elapsed >= timeoutSeconds) {
        return null;
      }

      await new Promise((resolve) => setTimeout(resolve, 500));
    }
  }

  async storeResult(
    taskId: string,
    result: Record<string, unknown>,
    status: string = 'success'
  ): Promise<void> {
    await this.connect();

    const message = this.createMessage(
      MessageType.RESULT,
      'orchestrator',
      { status, result }
    );
    message.id = taskId;

    const resultKey = `swarm:results:${taskId}`;
    await this.client!.setex(resultKey, 300, JSON.stringify(message));

    // Notificar orchestrator
    await this.client!.publish('swarm:results:orchestrator', JSON.stringify(message));

    // Remover de pendentes
    await this.client!.del(`swarm:pending:${taskId}`);
  }

  async broadcast(action: string, message: string = ''): Promise<string> {
    const msg = this.createMessage(
      MessageType.BROADCAST,
      '*',
      { action, message },
      Priority.HIGH
    );

    await this.publish('broadcast', msg);
    return msg.id;
  }

  async heartbeat(): Promise<void> {
    await this.connect();

    const key = `swarm:heartbeat:${config.agentId}`;
    const timestamp = Date.now() / 1000;
    await this.client!.setex(key, config.heartbeatTtl, timestamp.toString());
  }

  async healthCheck(): Promise<HealthCheckResult> {
    await this.connect();

    const workers: HealthCheckResult['workers'] = [];
    const keys = await this.client!.keys('swarm:heartbeat:*');
    const now = Date.now() / 1000;

    for (const key of keys) {
      const workerName = key.replace('swarm:heartbeat:', '');
      const lastSeen = await this.client!.get(key);

      let ageSeconds: number;
      let status: 'alive' | 'dead' | 'unknown';

      if (lastSeen) {
        ageSeconds = Math.floor(now - parseFloat(lastSeen));
        status = ageSeconds < 30 ? 'alive' : 'dead';
      } else {
        ageSeconds = -1;
        status = 'unknown';
      }

      workers.push({
        name: workerName,
        status,
        lastSeenSecondsAgo: ageSeconds,
      });
    }

    return {
      workers,
      total: workers.length,
      healthy: workers.filter((w) => w.status === 'alive').length,
      unhealthy: workers.filter((w) => w.status !== 'alive').length,
    };
  }

  async setState(
    key: string,
    value: unknown,
    ttlSeconds?: number
  ): Promise<void> {
    await this.connect();

    const fullKey = `swarm:state:${key}`;
    const valueStr = typeof value === 'string' ? value : JSON.stringify(value);

    if (ttlSeconds) {
      await this.client!.setex(fullKey, ttlSeconds, valueStr);
    } else {
      await this.client!.set(fullKey, valueStr);
    }
  }

  async getState(key: string): Promise<unknown | null> {
    await this.connect();

    const fullKey = `swarm:state:${key}`;
    const value = await this.client!.get(fullKey);

    if (value) {
      try {
        return JSON.parse(value);
      } catch {
        return value;
      }
    }
    return null;
  }
}

// Instancia global
export const broker = new MessageBroker();
